File: server/todo/Services/TodoService.cs
using todo.Dtos;
using todo.Enums;
using todo.Models;

namespace todo.Services;

public interface ITodoService
{
    void CreateTodoItem(CreateTodoItemDto todoItemDto);
    TodoItemDto GetTodoItemById(int id);
    void DeleteTodoItem(int testTodoId);
    void UpdateTodoItem(int testTodoId, UpdateTodoItemDto updateTodoItemDto);
    IEnumerable<TodoItemDto> GetTodoItems(TodoItemStatus? status, DateTime? dueDate, string sortBy, string sortDirection);

}

public class TodoService : ITodoService
{
    private readonly TodoContext _context;

    public TodoService(TodoContext context)
    {
        _context = context;
    }
    public void CreateTodoItem(CreateTodoItemDto todoItemDto)
    {
        var todoItem = new TodoItemDto()
        {
            Name = todoItemDto.Name,
            Description = todoItemDto.Description,
            DueDate = todoItemDto.DueDate,
            Status = TodoItemStatus.NotStarted
        };

        _context.TodoItems.Add(todoItem);

        _context.SaveChanges();
    }

    public TodoItemDto GetTodoItemById(int id)
    {
        var todoItem = _context.TodoItems.Find(id);

        if (todoItem == null)
        {
            throw new KeyNotFoundException($"Todo item with ID {id} not found.");
        }

        return todoItem;
    }

    public void DeleteTodoItem(int id)
    {
        var todoItem = GetTodoItemById(id);

       _context.TodoItems.Remove(todoItem);

       _context.SaveChanges();
    }

    public void UpdateTodoItem(int id, UpdateTodoItemDto updateTodoItemDto)
    {
        var todoItem = GetTodoItemById(id);
        todoItem.Status = updateTodoItemDto.Status;
        todoItem.Description = updateTodoItemDto.Description;
        todoItem.DueDate = updateTodoItemDto.DueDate;
        todoItem.Name = updateTodoItemDto.Name;

        _context.SaveChanges();
    }

    public IEnumerable<TodoItemDto> GetTodoItems(TodoItemStatus? status, DateTime? dueDate, string sortBy, string sortDirection)
    {
        var query = _context.TodoItems.AsQueryable();
        if (status != null)
        {
            query = query.Where(todo => todo.Status == status);
        }

        if (dueDate.HasValue)
        {
            query = query.Where(todo => todo.DueDate.Date == dueDate.Value.Date);
        }
        
        query = sortBy.ToLower() switch
        {
            "duedate" => sortDirection.ToLower() == "desc"
                ? query.OrderByDescending(todo => todo.DueDate)
                : query.OrderBy(todo => todo.DueDate),
            "status" => sortDirection.ToLower() == "desc"
                ? query.OrderByDescending(todo => todo.Status)
                : query.OrderBy(todo => todo.Status),
            "name" => sortDirection.ToLower() == "desc"
                ? query.OrderByDescending(todo => todo.Name)
                : query.OrderBy(todo => todo.Name),
            _ => query.OrderBy(todo => todo.Name) 
        };

        
        return query.Select(todo => new TodoItemDto
        {
            Id = todo.Id,
            Name = todo.Name,
            Description = todo.Description,
            DueDate = todo.DueDate,
            Status = todo.Status
        }).ToList();

    }
}

File: client/app/services/todo.service.ts
import axios from "axios";
import {
  TodoItemDto,
  CreateTodoItemDto,
  UpdateTodoItemDto,
  TodoItemStatus,
  Todo,
} from "../todo.models";

const API_BASE_URL = `${process.env.NEXT_PUBLIC_API_URL}/TodoItems`;

export const mapDTOtoModel = (dto: TodoItemDto) => {
  return new Todo(dto.id, dto.name, dto.description, dto.status, dto.dueDate);
};

export const getTodoItems = async (
  status?: TodoItemStatus,
  dueDate?: Date,
  sortBy: string = "Name",
  sortDirection: string = "asc"
) => {
  if (dueDate && !isDateValid(dueDate)) {
    dueDate = undefined;
  }
  try {
    const response = await axios.get<TodoItemDto[]>(`${API_BASE_URL}`, {
      params: {
        status,
        dueDate,
        sortBy,
        sortDirection,
      },
    });
    return response.data.map((x) => mapDTOtoModel(x));
  } catch (error) {
    console.error("Error fetching todo items", error);
    throw error;
  }
};

export const getTodoItem = async (id: number) => {
  try {
    const response = await axios.get<TodoItemDto>(`${API_BASE_URL}/${id}`);
    return response.data;
  } catch (error) {
    console.error("Error fetching todo item", error);
    throw error;
  }
};

export const createTodoItem = async (todoItemDto: CreateTodoItemDto) => {
  try {
    const response = await axios.post(`${API_BASE_URL}`, todoItemDto);
    return response.data;
  } catch (error) {
    console.error("Error creating todo item", error);
    throw error;
  }
};

export const updateTodoItem = async (
  id: number,
  todoItemDto: UpdateTodoItemDto
) => {
  try {
    await axios.put(`${API_BASE_URL}/${id}`, todoItemDto);
  } catch (error) {
    console.error("Error updating todo item", error);
    throw error;
  }
};

export const deleteTodoItem = async (id: number) => {
  try {
    await axios.delete(`${API_BASE_URL}/${id}`);
  } catch (error) {
    console.error("Error deleting todo item", error);
    throw error;
  }
};

const isDateValid = (date: Date) => {
  return !isNaN(date as any);
};
